@xen-orchestra/proxy/src/app/mixins/reverseProxy.mjs

```diff
@@ -0,0 +1,117 @@
+import { urlToHttpOptions } from 'url' // node  14.18.0, 15.7.0 is it ok ?
+import proxy from 'http2-proxy'
+
+function removeSlash(str) {
+  return str.replace(/^\/|\/$/g, '')
+}
+
+function mergeUrl(relative, base) {
+  const res = new URL(base)
+  const relativeUrl = new URL(relative, base)
+  res.pathname = relativeUrl.pathname
+  relativeUrl.searchParams.forEach((value, name) => {
+    // we do not allow to modify params already specified by config
+    if (!res.searchParams.has(name)) {
+      res.searchParams.append(name, value)
+    }
+  })
+  res.hash = relativeUrl.hash.length > 0 ? relativeUrl.hash : res.hash
+  return res
+}
+export default class ReverseProxy {
+  constructor(app, { httpServer }) {
+    this._app = app
+    httpServer.on('request', this.proxy)
+    httpServer.on('upgrade', this.upgrade)
+  }
+
+  localToBackendUrl(basePath, target, localPath) {
+    let localPathWithoutBase = removeSlash(localPath).substring(basePath.length)
+    localPathWithoutBase = './' + removeSlash(localPathWithoutBase)
+    const url = mergeUrl(localPathWithoutBase, target)
+    return url
+  }
+
+  backendToLocalPath(basePath, target, backendUrl) {
+    // keep redirect url relative to local server
+    const localPath = `${basePath}/${backendUrl.pathname.substring(target.pathname.length)}${backendUrl.search}${
+      backendUrl.hash
+    }`
+    return localPath
+  }
+
+  _getConfigFromRequest(req) {
+    for (const [path, config] of Object.entries(this._app.config.get('reverseProxies'))) {
```

Comment:
We should probably do this: https://github.com/vatesfr/xen-orchestra/blob/e9ca13aa12f030ae5a27b5a508373a8c20580eb2/packages/xo-server/src/index.mjs#L485

Reply:
I add a / at the end of the path to ensure that the proxy is exactly matched

Reply:
Sorting prefixes by descending length would allow advanced use like:

```toml
[reverseProxies]
'/foo/' = 'http://localhost:25469'
'/foo/bar' = 'http://localhost:37687'
```

Reply:
is that a prerequisite or should I add a todo and a warning in the doc ?

Reply:
It should be trivial to implement:

```suggestion
    const proxies = this._app.config.get('reverseProxies')
    for (const path of Object.keys(proxies).sort((a, b) => b.length - a.length)) {
      const config = proxies[path]
```